src/neo4j-schema/Neo4jSchemaTree.js

```diff
@@ -0,0 +1,171 @@
+  constructor(driver, config = {}) {
+    this.driver = driver;
+    this.nodes = {};
+    this.rels = {};
+  }
+
+  toJSON() {
+    return {
+      nodes: this.nodes,
+      rels: this.rels
+    };
+  }
+
+  initialize() {
+    const nodeTypeProperties = session =>
+      session
+        .run('CALL db.schema.nodeTypeProperties()')
+        .then(results => results.records.map(rec => rec.toObject()));
+
+    const relTypeProperties = session =>
+      session
+        .run('CALL db.schema.relTypeProperties()')
+        .then(results => results.records.map(rec => rec.toObject()));
+
+    console.log('Initializing your Neo4j Schema');
+    console.log('This may take a few moments depending on the size of your DB');
+    return Promise.all([
+      withSession(this.driver, nodeTypeProperties),
+      withSession(this.driver, relTypeProperties)
+    ])
+      .then(([nodeTypes, relTypes]) => this._populate(nodeTypes, relTypes))
+      .then(() => this._populateRelationshipLinkTypes())
+      .then(() => this);
+  }
+
+  _populateRelationshipLinkTypes() {
+    // console.log('Getting from/to relationship metadata');
+
+    const okapiIds = Object.keys(this.rels);
+
+    const promises = okapiIds.map(okapiId => {
+      const q = `
+                MATCH (n)-[r${okapiId}]->(m)
+                WITH n, r, m LIMIT 10
+                RETURN distinct(labels(n)) as from, labels(m) as to
+            `;
+
+      return withSession(this.driver, s =>
+        s.run(q).then(results => results.records.map(r => r.toObject()))
+      ).then(rows => {
+        this.getRel(okapiId).relType = extractRelationshipType(okapiId);
+        this.getRel(okapiId).links = rows;
+      });
+    });
+
+    return Promise.all(promises).then(() => this);
+  }
+
+  getNode(id) {
+    return this.nodes[id];
+  }
+
+  getNodes() {
+    return Object.values(this.nodes);
+  }
+
+  /**
+   * @param {Array[String]} labels a set of labels
+   * @returns {Neo4jNode} if it exists, null otherwise.
+   */
+  getNodeByLabels(labels) {
+    const lookingFor = _.uniq(labels);
+    const total = lookingFor.length;
+
+    return this.getNodes().filter(n => {
+      const here = n.getLabels();
+
+      const matches = here.filter(label => lookingFor.indexOf(label) > -1)
+        .length;
+      return matches === total;
+    })[0];
+  }
+
+  getRel(id) {
+    return this.rels[id];
+  }
+
+  getRels() {
+    return Object.values(this.rels);
+  }
+
+  _populate(nodeTypes, relTypes) {
+    // Process node types first
+    _.uniq(nodeTypes.map(n => n.nodeType)).forEach(nodeType => {
+      // A node type is an OKAPI node type label, looks like ":`Event`"
+      // Not terribly meaningful, but a grouping ID
+      const labelCombos = _.uniq(
+        nodeTypes.filter(i => i.nodeType === nodeType)
+      );
+
+      labelCombos.forEach(item => {
+        const combo = item.nodeLabels;
+        // A label combination is an array of strings ["X", "Y"] which indicates
```

Comment:
`@relation` fields for nodes with multiple labels seem to be duplicated in the typedefs. For example, I have a database with some nodes that have `:Session:Workshop`, which results in duplicate field definitions - in this case  `is_of_type`, `takes_place_in`, `speakers`, etc:

```

type Session_Workshop {
   _id: Long!
   date: Date!
   description: String!
   endtime: LocalTime!
   event_key: String!
   name: String!
   starttime: LocalTime!
   is_of_type: [Event_Type] @relation(name: "IS_OF_TYPE", direction: "OUT")
   takes_place_in: [Venue] @relation(name: "TAKES_PLACE_IN", direction: "OUT")
   has_tag: [Tag] @relation(name: "HAS_TAG", direction: "OUT")
   is_of_type: [Event_Type] @relation(name: "IS_OF_TYPE", direction: "OUT")
   takes_place_in: [Venue] @relation(name: "TAKES_PLACE_IN", direction: "OUT")
   has_tag: [Tag] @relation(name: "HAS_TAG", direction: "OUT")
   speakers: [Speaker] @relation(name: "SPEAKS_AT", direction: "IN")
   speakers: [Speaker] @relation(name: "SPEAKS_AT", direction: "IN")
}
```

Not sure where exactly in the code this case is handled, so dropping it here.

Reply:
I think I know how to fix this but I'd like a copy of your zipped graph.db so I can test against your multi-labeled dataset

Reply:
oooh this was a really good one.  I've pushed a fix and now this database works for me.  But this was a great one to have found.  I was generating relationships **per incident label type** and in the case of Session_Workshop you have 2 labels, so it was getting done twice.  

I didn't really have to change much code, I just removed an outer loop and made it possible to detect if a Neo4jRelationship is incident to a label set, and then it was all good.

Added unit tests for this case.